### ISSUE TITLE
cvise does not work on OpenBSD 6.9 on PowerPC G5 Mac

### ISSUE THREAD

**Issue by seanm** [marxin/cvise#68]
For fun I'm playing with OpenBSD 6.9 on an old PowerPC G5 Mac.

I successfully built cvise but a basic invocation crashes:

```
cvise --version            
Segmentation fault (core dumped) 
```

Here's the backtrace:

```
builder3$ gdb python3 /home/builder/external/cvise-install/bin/python3.8.core 

(gdb) bt
#0  0xe06cb15c in do_mkvalue () from /usr/local/lib/libpython3.8.so.0.0
#1  0xe06cb3ac in do_mktuple () from /usr/local/lib/libpython3.8.so.0.0
#2  0xe06ca9c0 in do_mkvalue () from /usr/local/lib/libpython3.8.so.0.0
#3  0xe06c9b08 in Py_BuildValue () from /usr/local/lib/libpython3.8.so.0.0
#4  0xe08fc1bc in psutil_proc_oneshot_info () from /usr/local/lib/python3.8/site-packages/psutil/_psutil_bsd.cpython-38.so
#5  0xe05744a0 in cfunction_call_varargs () from /usr/local/lib/libpython3.8.so.0.0
#6  0xe0575458 in PyCFunction_Call () from /usr/local/lib/libpython3.8.so.0.0
#7  0xe0573af8 in _PyObject_MakeTpCall () from /usr/local/lib/libpython3.8.so.0.0
#8  0xe068662c in call_function () from /usr/local/lib/libpython3.8.so.0.0
#9  0xe06836a0 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#10 0xe067d584 in PyEval_EvalFrameEx () from /usr/local/lib/libpython3.8.so.0.0
#11 0xe0574910 in function_code_fastcall () from /usr/local/lib/libpython3.8.so.0.0
#12 0xe0574ad4 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#13 0xe0686568 in call_function () from /usr/local/lib/libpython3.8.so.0.0
#14 0xe0683724 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#15 0xe0687424 in _PyEval_EvalCodeWithName () from /usr/local/lib/libpython3.8.so.0.0
#16 0xe0574b54 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#17 0xe0686568 in call_function () from /usr/local/lib/libpython3.8.so.0.0
#18 0xe0683684 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#19 0xe067d584 in PyEval_EvalFrameEx () from /usr/local/lib/libpython3.8.so.0.0
#20 0xe0574910 in function_code_fastcall () from /usr/local/lib/libpython3.8.so.0.0
#21 0xe0574ad4 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#22 0xe0574070 in PyVectorcall_Call () from /usr/local/lib/libpython3.8.so.0.0
#23 0xe05741fc in PyObject_Call () from /usr/local/lib/libpython3.8.so.0.0
#24 0xe0683948 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#25 0xe0687424 in _PyEval_EvalCodeWithName () from /usr/local/lib/libpython3.8.so.0.0
#26 0xe0574b54 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#27 0xe0686568 in call_function () from /usr/local/lib/libpython3.8.so.0.0
#28 0xe0683684 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#29 0xe067d584 in PyEval_EvalFrameEx () from /usr/local/lib/libpython3.8.so.0.0
#30 0xe0574910 in function_code_fastcall () from /usr/local/lib/libpython3.8.so.0.0
#31 0xe0574ad4 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#32 0xe0686568 in call_function () from /usr/local/lib/libpython3.8.so.0.0
#33 0xe0683684 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#34 0xe0687424 in _PyEval_EvalCodeWithName () from /usr/local/lib/libpython3.8.so.0.0
#35 0xe0574b54 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#36 0xe0686568 in call_function () from /usr/local/lib/libpython3.8.so.0.0
#37 0xe0683684 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#38 0xe0687424 in _PyEval_EvalCodeWithName () from /usr/local/lib/libpython3.8.so.0.0
#39 0xe0574b54 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#40 0xe05738bc in _PyObject_FastCallDict () from /usr/local/lib/libpython3.8.so.0.0
#41 0xe05759ac in _PyObject_Call_Prepend () from /usr/local/lib/libpython3.8.so.0.0
#42 0xe05fbeac in slot_tp_init () from /usr/local/lib/libpython3.8.so.0.0
#43 0xe05f5500 in type_call () from /usr/local/lib/libpython3.8.so.0.0
#44 0xe0573af8 in _PyObject_MakeTpCall () from /usr/local/lib/libpython3.8.so.0.0
#45 0xe068662c in call_function () from /usr/local/lib/libpython3.8.so.0.0
#46 0xe06836a0 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#47 0xe067d584 in PyEval_EvalFrameEx () from /usr/local/lib/libpython3.8.so.0.0
#48 0xe0574910 in function_code_fastcall () from /usr/local/lib/libpython3.8.so.0.0
#49 0xe0574ad4 in _PyFunction_Vectorcall () from /usr/local/lib/libpython3.8.so.0.0
#50 0xe0686568 in call_function () from /usr/local/lib/libpython3.8.so.0.0
#51 0xe0683724 in _PyEval_EvalFrameDefault () from /usr/local/lib/libpython3.8.so.0.0
#52 0xe0687424 in _PyEval_EvalCodeWithName () from /usr/local/lib/libpython3.8.so.0.0
#53 0xe067d358 in PyEval_EvalCode () from /usr/local/lib/libpython3.8.so.0.0
#54 0xe06dd0c0 in run_mod () from /usr/local/lib/libpython3.8.so.0.0
#55 0xe06dabb4 in PyRun_SimpleFileExFlags () from /usr/local/lib/libpython3.8.so.0.0
#56 0xe06da5f0 in PyRun_AnyFileExFlags () from /usr/local/lib/libpython3.8.so.0.0
#57 0xe07044f4 in Py_RunMain () from /usr/local/lib/libpython3.8.so.0.0
#58 0xe0704a14 in pymain_main () from /usr/local/lib/libpython3.8.so.0.0
#59 0xe0704ba4 in Py_BytesMain () from /usr/local/lib/libpython3.8.so.0.0
#60 0x249206b4 in main () from /usr/local/bin/python3
```

Hopefully that means something to you... otherwise, it does repro, so I could try whatever...

**Comment by marxin** [comment]
Hello.

Thank you for the interesting report. It seems to me that it's one of the following `psutil` calls:
```
psutil.cpu_count(logical=False)
psutil.cpu_count(logical=True)
psutil.Process().cpu_affinity()
```

What version of `psutil` do you have? Can you please run these individual calls in terminal?

**Comment by seanm** [comment]
Thanks for your reply.  Here you go:

```
builder3$ python3 
Python 3.8.8 (default, Apr 19 2021, 14:29:36) 
[Clang 10.0.1 ] on openbsd6
Type "help", "copyright", "credits" or "license" for more information.
>>> import psutil
>>> psutil.version_info
(5, 4, 3)
>>> psutil.cpu_count(logical=False)
>>> psutil.cpu_count(logical=True)
2
>>> psutil.Process().cpu_affinity()
Segmentation fault (core dumped) 
builder3$ 
```

**Comment by marxin** [comment]
All right, so we know the culprit method. The method is supported by FreeBSD:
https://psutil.readthedocs.io/en/latest/index.html?highlight=cpu_affinity#psutil.Process.cpu_affinity

It's a C implementation:
https://github.com/giampaolo/psutil/blob/master/psutil/arch/freebsd/specific.c#L810

Can you please create an issue at:
https://github.com/giampaolo/psutil/issues

**Comment by seanm** [comment]
It looks like it's not supported on OpenBSD according to the comments here: https://github.com/giampaolo/psutil/blob/master/psutil/_psutil_bsd.c

So probably this would happen on Intel too, and the old PowerPC G5 is not the issue.

A seg fault seems an odd way to not support something, as opposed to just doing nothing.

https://github.com/giampaolo/psutil/issues/2022

Do you think commenting out the affinity usage in cvise would work for me?  Guess I'll try...

**Comment by marxin** [comment]
> It looks like it's not supported on OpenBSD according to the comments here: https://github.com/giampaolo/psutil/blob/master/psutil/_psutil_bsd.c

I see!
 
> So probably this would happen on Intel too, and the old PowerPC G5 is not the issue.
> 
> A seg fault seems an odd way to not support something, as opposed to just doing nothing.

Sure! It should not crash.

> 
> Here's a ticket for psutil: [giampaolo/psutil#2022](https://github.com/giampaolo/psutil/issues/2022)
> 
> Do you think commenting out the affinity usage in cvise would work for me? Guess I'll try...

Yeah, it's not important, it's part of the automatic detection of CPU count. Just comment it out and let me know.

**Comment by seanm** [comment]
>Just comment it out and let me know

```
builder3$ /home/builder/external/cvise-install/bin/cvise --version 
cvise 2.4.0 (ad3e554)
```

Success!  Now I'll see if it can do real work...

**Comment by seanm** [comment]
So I hit another issue:

```
builder3$ /home/builder/external/cvise-install/bin/cvise ./test.sh small.c                                                                                                                 
C-Vise cannot run because the interestingness test does not return
zero. Please ensure that it does so not only in the directory where
you are invoking C-Vise, but also in an arbitrary temporary
directory containing only the files that are being reduced. In other
words, running these commands:

  DIR=`mktemp -d`
  cp /home/builder/external/clang52438/small.c $DIR
  cd $DIR
  /home/builder/external/clang52438/test.sh
  echo $?

should result in '0' being echoed to the terminal.
Please ensure that the test script takes no arguments; it should be hard-coded to refer
to the same file that is passed as an argument to C-Vise.

See 'cvise.py --help' for more information.
```

Then I tried the commands it says:

```
builder3$ DIR=`mktemp -d`
builder3$ cp /home/builder/external/clang52438/small.c $DIR
builder3$ cd $DIR
builder3$ pwd
/tmp/tmp.Lp6WHErYWE
builder3$ /home/builder/external/clang52438/test.sh
ksh: /home/builder/external/clang52438/test.sh: No such file or directory
```

yet the files does exist:

```
builder3$ file /home/builder/external/clang52438/test.sh
/home/builder/external/clang52438/test.sh: Bourne-Again shell script text executable
```

But if I run it this way:

```
builder3$ sh /home/builder/external/clang52438/test.sh
builder3$ echo $?
0
```

I think this may be a bash vs other shell thing...

**Comment by marxin** [comment]
> ```
> builder3$ /home/builder/external/clang52438/test.sh
> ksh: /home/builder/external/clang52438/test.sh: No such file or directory
> ```

This is super weird, why does it say that the file does not exist? Can you please switch to `bash` shell for instance in order to run the generated `.sh` file. What about running `ksh /home/builder/external/clang52438/test.sh`?

**Comment by marxin** [comment]
Hm, that's weird, I see:
```
cvise ./test.sh small.c
00:00:00 INFO ===< 341 >===
00:00:00 INFO running 8 interestingness tests in parallel
00:00:00 INFO INITIAL PASSES
00:00:00 INFO ===< IncludesPass >===
00:00:00 INFO ===< UnIfDefPass >===
00:00:00 INFO ===< CommentsPass >===
00:00:00 INFO ===< IfPass >===
00:00:00 INFO ===< LineMarkersPass >===
00:00:00 INFO ===< BlankPass >===
00:00:00 INFO (3.0%, 64 bytes, 8 lines)
00:00:00 INFO ===< ClangBinarySearchPass::replace-function-def-with-decl >===
00:00:00 INFO using C++ standard: c++98 with 1 transformation opportunities
00:00:00 INFO ===< ClangBinarySearchPass::remove-unused-function >===
00:00:01 INFO using C++ standard: c++98 with 0 transformation opportunities
00:00:01 INFO ===< LinesPass::0 >===
00:00:01 INFO ===< LinesPass::1 >===
00:00:01 INFO (51.5%, 32 bytes, 3 lines)
00:00:01 INFO (77.3%, 15 bytes, 2 lines)
00:00:01 INFO (81.8%, 12 bytes, 1 lines)
...
```

Please try using `--debug` option.

**Comment by seanm** [comment]
```
builder3$ /home/builder/external/cvise-install/bin/cvise --debug ./test.sh small.c
00:00:00 DEBUG perform sanity check... 
00:00:00 DEBUG sanity check tmpdir = /tmp/cvise-sanity-1mhd6wg9
00:00:00 DEBUG sanity check successful
00:00:00 INFO ===< 90522 >===
00:00:00 INFO running 2 interestingness tests in parallel
00:00:00 INFO INITIAL PASSES
00:00:00 DEBUG Creating pass root folder: /tmp/cvise-IncludesPass-_nn33m7e
00:00:00 INFO ===< IncludesPass >===
```

It's been "only" 30 minutes, but I think it's stuck.  `top` says CPU is 99% idle.

**Comment by marxin** [comment]
Hmm, hard to guess. Can you attach gdb to the python process and print backtrace, please?
It may be possible that `Pebble` library (that I use for process dispatching) does not work on BSD.

**Comment by seanm** [comment]
To eliminate PowerPC being the issue, I've created an OpenBSD 7 VM on Intel and am trying there.  I forget if I hit this on PowerPC, but I'm getting this build error:

```
[  5%] Building CXX object clang_delta/CMakeFiles/clang_delta.dir/TransformationManager.cpp.o
In file included from /home/sean/external/cvise/clang_delta/TransformationManager.cpp:25:
In file included from /usr/local/include/clang/Frontend/CompilerInstance.h:15:
In file included from /usr/local/include/clang/Frontend/CompilerInvocation.h:22:
/usr/local/include/clang/StaticAnalyzer/Core/AnalyzerOptions.h:437:9: error: 'PACKAGE' macro redefined [-Werror,-Wmacro-redefined]
#define PACKAGE(FULLNAME) llvm::StringLiteral(FULLNAME),
        ^
/home/sean/external/cvise-bin/config.h:35:9: note: previous definition is here
#define PACKAGE "cvise"
        ^
1 error generated.
```

IIRC OpenBSD sometimes defaults to using `-Werror`... I got past it by specifying `-w` in my C/C++ flags.  But maybe you could fix this warning?

**Comment by marxin** [comment]
> /usr/local/include/clang/StaticAnalyzer/Core/AnalyzerOptions.h:437:9: error: 'PACKAGE' macro redefined [-Werror,-Wmacro-redefined]
> #define PACKAGE(FULLNAME) llvm::StringLiteral(FULLNAME),

This comes from LLVM itself, what version do you use? What compiler do you use for the compilation of C-Vise?
You remove the `-Werror` in `CMakeLists.txt`.

**Comment by seanm** [comment]
I'm using the clang that comes with OpenBSD:

```
$ clang --version
OpenBSD clang version 11.1.0
Target: amd64-unknown-openbsd7.0
Thread model: posix
InstalledDir: /usr/bin
```

and the LLVM libs that cvise needed to build are from `pkg_add llvm` (which gives llvm-11.1.0p7).

Yes, AnalyzerOptions.h:437 is from LLVM, but the other define is from cvise/cmake_config.h.in I think.  Is it meant to be the same `PACKAGE` macro, or is this a name collision?

**Comment by marxin** [comment]
All right, I fixed that on `master`, please pull.

**Comment by seanm** [comment]
Cool, thanks!

>It may be possible that Pebble library...

In fact, I think I didn't have pebble installed, or I installed it with `pkg_add`, not `pip`.  I have deleted all packages on the PowerPC and will now try fresh...

**Comment by seanm** [comment]
Success!  I can now reduce a toy example correctly!

For anyone else trying on OpenBSD, here's what worked for me:

```
doas pkg_add python3 py3-pip llvm bash cmake git
doas ln -sf /usr/local/bin/pip3.8 /usr/local/bin/pip

python3 -m pip install --upgrade psutil
python3 -m pip install --upgrade Pytest
python3 -m pip install --upgrade Pebble
python3 -m pip install --upgrade chardet
python3 -m pip install --upgrade flake8

git clone https://github.com/marxin/cvise.git
mkdir cvise-bin
mkdir cvise-install
cd cvise-bin
ccmake ../cvise
(set install folder, set to build Release because Debug fails to link)
(hack cvise to remove affinity stuff)
make -j2 install
```

So the only issue remaining is the affinity thing, which is not a cvise bug really.

**Comment by marxin** [comment]
Good, thank you for your help. Lemme close this.
